File: app/infra/py_infra/components/grpc_client/factory.py
# infra/pyinfra/components/grpc_client/factory.py
from typing import Dict
from components.grpc_client.component import GRPCClientComponent
from components.grpc_client.schema import GRPCClientsConfig, GRPCClientConfig
from core.container import Container
import logging

logger = logging.getLogger(__name__)


def create_and_register_grpc_clients(config: Dict) -> GRPCClientComponent:
    """创建并注册GRPC客户端组件"""

    # 解析配置
    grpc_config_dict = config.get("grpc_clients", {})

    # 转换客户端配置
    clients_config = {}
    for name, client_config in grpc_config_dict.get("clients", {}).items():
        clients_config[name] = GRPCClientConfig(**client_config, name=name)

    # 创建主配置
    grpc_clients_config = GRPCClientsConfig(
        clients=clients_config,
        **{k: v for k, v in grpc_config_dict.items() if k != "clients"}
    )

    # 创建组件
    component = GRPCClientComponent(grpc_clients_config)

    # 注册到容器
    Container.register("grpc_clients", component)

    # 注册到生命周期管理
    from core.lifecycle import LifecycleManager
    LifecycleManager.register_component(component)

    logger.info(f"GRPC clients component registered with {len(clients_config)} clients")
    return component

File: app/infra/go/application/components/grpc_client/factory.go
// app/infra/go/application/components/grpc_client/factory.go
package grpc_client

import (
	"fmt"
	"time"

	"github.com/grand-thief-cash/chaos/app/infra/go/application/components/logging"
	"github.com/grand-thief-cash/chaos/app/infra/go/application/core"
)

type Factory struct {
	logger logging.Logger
}

func NewFactory(logger logging.Logger) *Factory {
	return &Factory{logger: logger}
}

func (f *Factory) Create(cfg interface{}) (core.Component, error) {
	grpcConfig, ok := cfg.(*GRPCClientsConfig)
	if !ok {
		return nil, fmt.Errorf("invalid config type for GRPC clients component, expected *GRPCClientsConfig")
	}
	if !grpcConfig.Enabled {
		return nil, fmt.Errorf("GRPC clients component is disabled")
	}
	f.setDefaults(grpcConfig)
	return NewGRPCClientComponent(grpcConfig, f.logger), nil
}

func (f *Factory) setDefaults(cfg *GRPCClientsConfig) {
	if cfg.DefaultTimeout == 0 {
		cfg.DefaultTimeout = 30 * time.Second
	}
	if cfg.HealthCheckInterval == 0 {
		cfg.HealthCheckInterval = 60 * time.Second
	}
	for name, clientConfig := range cfg.Clients {
		if clientConfig.Name == "" {
			clientConfig.Name = name
		}
		if clientConfig.MaxReceiveMessageLength == 0 {
			clientConfig.MaxReceiveMessageLength = 4 * 1024 * 1024
		}
		if clientConfig.MaxSendMessageLength == 0 {
			clientConfig.MaxSendMessageLength = 4 * 1024 * 1024
		}
		if clientConfig.Timeout == 0 {
			clientConfig.Timeout = cfg.DefaultTimeout
		}
	}
}
